kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt

```diff
@@ -59,46 +72,67 @@ internal class WorkflowRunnerViewModel<OutputT : Any>(
       val hostFactory = WorkflowHost.Factory(dispatcher)
       val workflowHost = hostFactory.run(workflow, inputs, snapshot)
       @Suppress("UNCHECKED_CAST")
-      return WorkflowRunnerViewModel(viewRegistry, workflowHost) as T
+      return WorkflowRunnerViewModel(viewRegistry, workflowHost, dispatcher) as T
     }
   }
 
-  private val subs = CompositeDisposable()
+  private val snapshotJob = GlobalScope.launch(context) {
+    workflowHost.renderingsAndSnapshots
+        .map { it.snapshot }
+        .collect { lastSnapshot = it }
+  }
 
-  @Suppress("EXPERIMENTAL_API_USAGE")
-  private val updates =
-    workflowUpdates.updates.asObservable(Dispatchers.Unconfined)
-        .doOnNext { lastSnapshot = it.snapshot }
-        .replay(1)
-        .autoConnect(1) { subs.add(it) }
+  private var workflowJob: Job? = null
 
   private var lastSnapshot: Snapshot = Snapshot.EMPTY
 
-  override val renderings: Observable<out Any> = updates.map { it.rendering }
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  override val renderings: Observable<out Any> = workflowHost.renderingsAndSnapshots
+      // Lazily start the workflow the first time someone starts listening. Like autoConnect(1).
+      .onCollect {
+        // This call is idempotent and will always return the same Job, so we can just call it
+        // every time.
+        workflowJob = workflowHost.start()
+      }
+      .map { it.rendering }
+      .asObservable()
 
-  override val output: Flowable<out OutputT> =
-    // Buffer on backpressure so outputs don't get lost.
-    updates.toFlowable(BUFFER)
-        .filter { it.output != null }
-        .map { it.output!! }
-        // DON'T replay, outputs are events.
-        .publish()
-        // Subscribe upstream immediately so we immediately start getting notified about outputs.
-        // If [renderings] triggers the upstream subscription before we do, if the workflow emits
-        // an output immediately we might not see it.
-        .autoConnect(0) { subs.add(it) }
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  override val output: Flowable<out OutputT> = workflowHost.outputs
+      // Lazily start the workflow the first time someone starts listening. Like autoConnect(1).
```

Comment:
I found this comment more confusing than helpful -- at first it made me think you meant that `onCollect` is analogous to `autoConnect(1)`, but really it's just `onSubscribe`, right? We get the `autoConnect(1)` behavior by virtue of calling `start()`, and because `start()` is idempotent. And that seems well enough documented by the comment below. If I have this right then perhaps delete this comment and change the one below to:

```
  // This call is idempotent and will always return the same Job, so we can just call it
  // every time. And so by making this call inside of onCollect(), we basically get the
  // nice effects of Rx's autoConnect(1) start up a shared stream the the first time
  // someone "subscribes" / collects.
```

And obv. the same for the identical block above.

Reply:
Done (extracted into a function too).